CodingConventions.md

```diff
@@ -0,0 +1,307 @@
+     ...
+  }
+  ```
+* Handle errors at a level, where you still have enough context to decide whether the error is expected during normal
+  operations.
+* Errors of unexpected types are indicators that the node's internal state might be corrupted.
+
+### Anti-Pattern
+
+Continuing on a best-effort basis is not an option, i.e. the following is an anti-pattern in the context of Flow:
+
+```golang
+err := foo()
+if err != nil {
+   log.Error().Err(err).Msg("foo failed")
+   return
+}
+```
+
+There are _rare_ instances, where this might be acceptable. For example, when attempting to send a message via the
+network to another node and the networking layer errored. In this case, please include a comment with an explanation,
+why it is acceptable to keep going, even if the other component returned an error. For this example, it is acceptable to handle the error by logging because:
+* we expect transient errors from this component (networking layer) during normal operation
+* the networking layer uses a 3rd party library, which does not expose specific and exhaustive sentinel errors for expected failure conditions
+* either it is not critical that the message is successfully sent at all, or it will be retried later on
+
+### Prioritize Safety Over Liveness
+
+**Ideally, a vertex should restart** (from a known good state), **when it encounters an unexpected error.**
+Per convention, a vertex should throw any unexpected exceptions using the related [irrecoverable context](https://github.com/onflow/flow-go/blob/277b6515add6136946913747efebd508f0419a25/module/irrecoverable/irrecoverable.go).
+
+[TODO: expend this section] 
+
+If this is out of scope, we _prioritize safety over liveness_. This means that we rather crash the node than continue on a best-effort basis.
+When in doubt, use the following as a fall-back:
+```golang
+err := foo()
+if errors.Is(err, XFailedErr) {
+   // expected error
+   return
+}
+if err != nil {
+   log.Fatal().Err(err).Msg("unexpected internal error")
+   return
+}
+```
+
+The error-handling guidelines are set up such that they minimize technical debt, even if we don't support component
+restarts right away.
+
+* The code is already differentiating expected from unexpected errors in the core business logic and we documented them
+  accordingly.
+* The code handles the expected errors at the appropriate level, only propagating unexpected errors.
+
+As a first step, you can crash the node when an unexpected error hits the `Engine` level (for more details, see `Engine`
+section of this doc). If you have followed the error handling guidelines, leaving the restart logic for the component as
+technical debt for later hopefully does not add much overhead. In contrast, if you log errors and continue on a
+best-effort basis, you also leave the _entire_ logic for differentiating errors as tech debt. When adding this logic
+later, you need to revisit the _entire_ business logic again.
+
+
+# Appendix
+
+## Engines (interface deprecated)
```

Comment:
I think it would be helpful to explain why engines are deprecated since they are still so pervasive. Also, is there a plan to fix this - eg, a tracking GH issue?

Reply:
Linked a tracking issue with an explanation